crates/jstzd/src/protocol/mod.rs

```diff
@@ -0,0 +1,186 @@
+        let raw_json: Value = match &self.source_path {
+            Some(path) => {
+                let mut buffer = String::new();
+                match std::fs::File::open(path) {
+                    Ok(mut f) => {
+                        f.read_to_string(&mut buffer).unwrap();
+                        serde_json::from_slice(buffer.as_bytes())?
+                    }
+                    Err(e) => {
+                        return Err(anyhow::anyhow!(
+                            "Failed to open parameter file at {:?}: {}",
+                            path,
+                            e.to_string()
+                        ))
+                    }
+                }
+            }
+            None => {
+                let file_path = &self.protocol.parameter_file(&self.constants);
+                let file_path_str = file_path.to_str().ok_or(anyhow::anyhow!(
+                    "Failed to convert parameter file path to string"
+                ))?;
+                let f =
+                    ProtocolParameterFile::get(file_path_str).ok_or(anyhow::anyhow!(
+                        "Failed to load parameter file at '{}'",
+                        file_path_str
+                    ))?;
+                serde_json::from_slice(&f.data)?
+            }
+        };
+        Ok(raw_json)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{Protocol, ProtocolConstants, ProtocolParameterBuilder};
+
+    #[test]
+    fn parameter_builder() {
+        let mut builder = ProtocolParameterBuilder::new();
+        builder
+            .set_constants(ProtocolConstants::TestConstants)
+            .set_protocol(Protocol::TestVersion)
+            .set_source_path("/test/path");
+        assert_eq!(builder.constants, ProtocolConstants::TestConstants);
+        assert_eq!(
+            builder.source_path.unwrap().as_os_str().to_str().unwrap(),
+            "/test/path"
+        );
+        assert_eq!(builder.protocol.hash(), Protocol::TestVersion.hash());
+    }
+
+    #[test]
+    fn parameter_builder_default() {
+        let builder = ProtocolParameterBuilder::new();
+        assert_eq!(
+            builder.constants.to_string(),
+            ProtocolConstants::Sandbox.to_string(),
+        );
+        assert!(builder.source_path.is_none());
+        assert_eq!(builder.protocol, Protocol::Alpha);
+    }
+
+    #[test]
+    fn build_parameters_from_given_file() {
+        let mut builder = ProtocolParameterBuilder::new();
+        let source_path = format!(
+            "{}/resources/test/protocol_parameters/sandbox/test-hash",
+            env!("CARGO_MANIFEST_DIR")
+        );
+        builder.set_source_path(&source_path);
+        let output_path = builder.build().unwrap();
+        let file = std::fs::File::open(output_path).unwrap();
+        let json: serde_json::Value = serde_json::from_reader(file).unwrap();
+
+        // The given parameter file `test-hash` has only one kv-pair "foo:bar"
```

Comment:
Nit: I would have preferred to write the contents to a temp file since the contents of the file is tiny and we can easily see what to expect instead of commenting here.

Reply:
It is written to a temp file. `json` is loaded from the path to the temp file. I'll update the comment to avoid confusion.